Online distributed graph partitioning using PT-Scotch

abishekg7 opened the pull request:
This PR enables the use of MPAS atmosphere core without requiring any graph decomposition files specified as input, if the atmosphere core has been built with the Scotch graph partitioning library (built separately). 

This PR supersedes https://github.com/MPAS-Dev/MPAS-Model/pull/1348 to implement a Fortran interface to the PT-Scotch C library within MPAS. 

### Building MPAS with Scotch

Instructions to build Scotch are provided later in the description. The Scotch version must be at least v7.0.8. Once Scotch has been installed, the MPAS atmosphere core can be leveraged to use online graph partitioning by setting the path to Scotch prior to building MPAS.

```
export SCOTCH=/path/to/scotch/installation
make nvhpc CORE=atmosphere
```

If MPAS is built/linked with Scotch successfully, you should see a message stating that at the end of the build output. 

### Usage and run-time behavior

After building MPAS with Scotch, you can still choose whether or not to use online graph partitioning by setting appropriate values for the `config_block_decomp_file_prefix` namelist option.  

- If `config_block_decomp_file_prefix`+`mpi_tasks` points to a valid graph decomp file that already exists in the run directory, then it is used to proceed with the model run without invoking Scotch. 

- If `config_block_decomp_file_prefix==''` or `config_block_decomp_file_prefix`+`mpi_tasks` does not match any valid graph decomp file in the run directory, then Scotch graph partitioning is invoked. During this process, the generated partition is saved as a graph decomp file to the run directory so that it may be reused in the following runs without needing to invoke Scotch again. 

If MPAS has not been built with scotch, any code paths relating to Scotch will not be taken. And an incorrect specification of `config_block_decomp_file_prefix`+`mpi_tasks` should lead to the model halting.

### Downloading and building Scotch

Scotch may be obtained from its gitlab repository or as tarballs from this [link](https://gitlab.inria.fr/scotch/scotch/-/releases). The minimum Scotch version must be at least v7.0.8. 
```
git clone https://gitlab.inria.fr/scotch/scotch.git && cd scotch 
git checkout v7.0.8
```

Build instructions are on this [page](https://gitlab.inria.fr/scotch/scotch#installation)

It can be as simple as 
```
mkdir build && cd build
cmake -DCMAKE_INSTALL_PREFIX=prefix/scotch/install -DMPI_HOME=/your/mpi/path  ..
make -j5
make install
```

To be able to use the distributed graph partitioning provided by PT-Scotch, you will need to pass an appropriate MPI installation path during the Scotch build. 

Other system pre-requisites are Bison and Flex. I have observed that an older Flex version (v2.6.1) on Eris caused the Scotch build to fail. After pulling in a more recent version of Flex [v2.6.4](https://github.com/westes/flex/releases) it seemed to build fine.  Scotch documentation also references the requirement for a Bison version > 3.4.

mgduda commented on src/core_atmosphere/Registry.xml:
```diff
@@ -319,7 +319,7 @@
 
         <nml_record name="decomposition" in_defaults="true">
-                <nml_option name="config_block_decomp_file_prefix" type="character" default_value="x1.40962.graph.info.part."
+                <nml_option name="config_block_decomp_file_prefix" type="character" default_value=""
```
At least initially, I would guess that most people would not be linking with the PT-Scotch library, in which case it might be best to leave a the placeholder prefix "x1.40962.graph.info.part." here as a remind to set the prefix of an existing graph partition file. Those who read the documentation and understand how the online partitioning works can simply delete the placeholder prefix.

mgduda commented on Makefile:
```diff
@@ -1415,6 +1425,33 @@ musica_fortran_test:
+	@#
+	@# Create a Fortran test program that will link against the SCOTCH library
+	@#
+	$(info Checking for a working MUSICA-Fortran library...)
```
I think "MUSICA-Fortran" should be "Scotch" (or "PT-Scotch").

abishekg7 replied:
Fixed

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,304 @@
+
+contains
+
+    function scotch_dgraphinit(dgraph, comm) result(ierr)
```
Eventually, I think we'll want to have documentation blocks for each new function or subroutine.

abishekg7 replied:
Added some basic documentation blocks, but will double check the documentation.

mgduda commented on src/framework/mpas_bootstrapping.F:
```diff
@@ -430,6 +432,8 @@ subroutine mpas_bootstrap_framework_phase1(domain, mesh_filename, mesh_iotype, p
       deallocate(block_count)
       deallocate(readingBlock)
 
+      call mpas_timer_stop('bootstrap_framework_phase1')
```
Would it make sense to omit these timers as part of this PR, and instead, if we need more detailed timing of the bootstrapping, we could make a separate PR to add timers throughout?

abishekg7 replied:
Makes sense. I've removed these timers.

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,304 @@
+        implicit none
+
+        doubleprecision, target, intent(in) :: dgraph (SCOTCH_DGRAPHDIM)
+        integer(SCOTCH_NUMSIZE), intent(in) :: nVertices
```
I'm getting a compilation error related to `SCOTCH_NUMSIZE`:
```
mpas_ptscotch_interface.F(53): error #6683: A kind type parameter must be a compile-time constant.   [SCOTCH_NUMSIZE]
        integer(SCOTCH_NUMSIZE), intent(in) :: nVertices
----------------^
```
As far as I can tell, I've installed Scotch correctly (and I was able to use this library to test earlier versions of this PR). Is `SCOTCH_NUMSIZE` supposed to be defined in `ptscotchf.h`?

mgduda replied:
Or is there an option that I need to activate when compiling Scotch?

abishekg7 replied:
It seems like `SCOTCH_NUMSIZE` may have been added to `ptscotchf.h` in v7.0.8. Prior to that it was only present in `scotchf.h`. 

I could just go ahead and remove the reference to `SCOTCH_NUMSIZE`  as it's not technically needed in our case. As we don't plan on supporting Scotch builds with 64-bit integers. 

Alternatively, we could enforce Scotch >= 7.0.8.

mgduda replied:
I've updated to Scotch 7.0.10, and compilation is now successful!

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,412 @@
+#ifdef MPAS_SCOTCH
```
It would be good to add a copyright statement at the beginning of this file.

abishekg7 replied:
Fixed

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -195,7 +232,7 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
            close(unit=iunit)
            call mpas_release_unit(iunit)
-
+           
```
Let's undo this whitespace addition.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +322,239 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      vertloctab(partial_global_graph_info % nVertices+1) = nLocEdgesGraph + 1
+
+      ! Duplicate the communicator to be used by Scotch
+      call MPI_Comm_dup(MPI_COMM_WORLD, localcomm, mpi_ierr)
```
In general, we can't assume that MPAS owns the entire `MPI_COMM_WORLD` communicator (as might be the case, for example, when MPAS-Atmosphere is operating as a dynamical core in CAM within CESM).
Instead of `MPI_COMM_WORLD`, we can use the communicator found in `domain % dminfo`.

abishekg7 replied:
Makes sense. Thanks for the catch.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +322,239 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      if (dminfo % my_proc_id == IO_NODE) then
+         partitionFilePrefix=trim(blockFilePrefix)
+         if (trim(partitionFilePrefix) == '') then
+            write(partitionFilePrefix,'(a,i0,a)') 'x1.',partial_global_graph_info%nVerticesTotal,'.graph.info.part.'
```
I think we had decided not to use the `x1.` prefix here.

abishekg7 replied:
Fixed

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,238 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+
+      use mpas_timer, only : mpas_timer_start, mpas_timer_stop
+#ifdef MPAS_USE_MPI_F08
+      use mpi_f08, only : MPI_Comm, MPI_COMM_WORLD, MPI_INTEGER, MPI_Comm_dup, MPI_Comm_free, MPI_Gather, MPI_Gatherv
```
It doesn't look like `MPI_COMM_WORLD` is needed in this `use` statement.

abishekg7 replied:
Done

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,238 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      integer, dimension(:), allocatable :: local_block_list
+
+      integer, dimension(:), allocatable ::global_block_id_arr, local_block_id_arr
+      integer :: i, global_block_id, local_block_id, iunit, ounit, istatus, ostatus, j, k
```
With the `-Wall` flag, the gfortran compiler notes some unused variables:
```
mpas_block_decomp.F:375:35:

  375 |       integer :: i, global_block_id, local_block_id, iunit, ounit, istatus, ostatus, j, k
      |                                   1
Warning: Unused variable 'global_block_id' declared at (1) [-Wunused-variable]
mpas_block_decomp.F:375:74:

  375 |       integer :: i, global_block_id, local_block_id, iunit, ounit, istatus, ostatus, j, k
      |                                                                          1
Warning: Unused variable 'istatus' declared at (1) [-Wunused-variable]
mpas_block_decomp.F:375:58:

  375 |       integer :: i, global_block_id, local_block_id, iunit, ounit, istatus, ostatus, j, k
      |                                                          1
Warning: Unused variable 'iunit' declared at (1) [-Wunused-variable]
mpas_block_decomp.F:375:51:

  375 |       integer :: i, global_block_id, local_block_id, iunit, ounit, istatus, ostatus, j, k
      |                                                   1
Warning: Unused variable 'local_block_id' declared at (1) [-Wunused-variable]
mpas_block_decomp.F:371:59:

  371 |       integer, dimension(:), allocatable :: local_cell_list
      |                                                           1
Warning: Unused variable 'local_cell_list' declared at (1) [-Wunused-variable]
mpas_block_decomp.F:379:36:

  379 |       character (len=StrKIND) :: msg
      |                                    1
Warning: Unused variable 'msg' declared at (1) [-Wunused-variable]
mpas_block_decomp.F:385:54:

  385 |       doubleprecision :: scotchgraph (SCOTCH_GRAPHDIM)
      |                                                      1
Warning: Unused variable 'scotchgraph' declared at (1) [-Wunused-variable]
```

abishekg7 replied:
Thanks for the catch. Fixed.

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,285 @@
+{
+	SCOTCH_stratExit((SCOTCH_Strat *)strat_ptr);
+}
+#endif
```
It looks like this file is missing and EOF character.

abishekg7 replied:
Done.

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,418 @@
+    end subroutine scotch_dgraphdata
+
+end module mpas_ptscotch_interface
+#endif
```
Could you add an EOF character to this file?

abishekg7 replied:
Done.

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,418 @@
+        integer(SCOTCH_NUMSIZE), intent(out) :: local_cell_list(*)
+
+        interface
+            subroutine scotchfdgraphdata(dgraph_ptr, cell_list) bind(C, name='scotchm_dgraphdata')
```
Here and elsewhere in this file, would it make the code easier to follow if we declared these interface names to match the C names? Or is there a reason for using interface names that differ from the names of the C routines that we're calling?

abishekg7 replied:
I wasn't entirely sure if that was allowed. But I've fixed this now.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,238 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+
+      ! Duplicate the communicator to be used by Scotch
+      call MPI_Comm_dup(dminfo % comm, localcomm, mpi_ierr)
+      if (mpi_ierr .ne. 0) then
```
Just for consistency with other code in MPAS, could you change `.ne.` to `/=`?

abishekg7 replied:
Done.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,238 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      endif
+      ! Initialize the Scotch graph data structure, and an extra one to hold the re-distributed graph
+#ifdef MPAS_USE_MPI_F08
+      call scotch_dgraphinit(scotchdgraph(1), localcomm% mpi_val)
```
Here and in other calls, it would be more in line with the rest of MPAS to pass array arguments simply using their names rather than by passing their first element.

abishekg7 replied:
Done

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,418 @@
+
+        implicit none
+        ! Arguments
+        doubleprecision, target, intent(in) :: dgraph (SCOTCH_DGRAPHDIM)
```
Stylistically, I think it would be preferable to declare arrays using the `dimension` attribute. E.g.,
```fortran
doubleprecision, dimension(SCOTCH_DGRAPHDIM), target, intent(in) :: dgraph
```

abishekg7 replied:
Fixed.

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,285 @@
+void scotchm_dgraphdata(void *ptr, SCOTCH_Num *cell_list)
+{
+	
+	int err;
```
The `err` variable doesn't appear to be used.

abishekg7 replied:
Fixed.

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,285 @@
+	SCOTCH_Num *vendloctab = vertloctab_1 + 1;
+	SCOTCH_Num *edgeloctab = (SCOTCH_Num *)adjncy;
+
+	int i, err;
```
The `i` variable doesn't appear to be used anywhere in this function.

abishekg7 replied:
Fixed.

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,285 @@
+{
+	MPI_Comm comm;
+	MPI_Comm comm2;
+	int size, rank, err;
```
With the `-Wall` flag, the gcc compiler reports several unused variables:
```
ptscotch_interface.c: In function 'scotchm_dgraphinit':
ptscotch_interface.c:37:19: warning: unused variable 'rank' [-Wunused-variable]
   37 |         int size, rank, err;
      |                   ^~~~
ptscotch_interface.c:37:13: warning: unused variable 'size' [-Wunused-variable]
   37 |         int size, rank, err;
      |             ^~~~
ptscotch_interface.c:36:18: warning: unused variable 'comm2' [-Wunused-variable]
   36 |         MPI_Comm comm2;
      |                  ^~~~~
```

abishekg7 replied:
Fixed.

mgduda commented on src/framework/Makefile:
```diff
@@ -85,7 +87,7 @@ mpas_timekeeping.o: mpas_string_utils.o mpas_kind_types.o mpas_derived_types.o m
 mpas_timer.o: mpas_kind_types.o mpas_dmpar.o mpas_threading.o mpas_log.o
 
-mpas_block_decomp.o: mpas_derived_types.o mpas_hash.o mpas_io_units.o mpas_dmpar.o
+mpas_block_decomp.o: mpas_derived_types.o mpas_hash.o mpas_io_units.o mpas_dmpar.o mpas_ptscotch_interface.o
```
I think `mpas_block_decomp.o` now depends on `mpas_timer.o` as well.

mgduda commented on src/framework/Makefile:
```diff
@@ -112,6 +114,8 @@ xml_stream_parser.o: xml_stream_parser.c
 
 mpas_halo.o: mpas_derived_types.o mpas_pool_routines.o mpas_log.o
 
+mpas_ptscotch_interface.o : mpas_derived_types.o mpas_dmpar.o mpas_log.o ptscotch_interface.o
```
I doesn't look like the current version of the `mpas_ptsctoch_interface` depends on `mpas_dmpar`.

Also, although `mpas_ptscotch_interface` makes calls to routines that are implemented in `ptscotch_interface`, I think we can drop the dependency on `ptscotch_interface.o` since compilation of `ptscotch_interface` doesn't produce any artifacts (e.g., a `.mod` file) that are needed to compile `mpas_ptscotch_interface`.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,240 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      ! Duplicate the communicator to be used by Scotch
+      call MPI_Comm_dup(dminfo % comm, localcomm, mpi_ierr)
+      if (mpi_ierr /= 0) then
+            call mpas_log_write('Cannot duplicate communicator')
```
Would it be better to make this a critical error? It seems like not being able to duplicate `dminfo % comm` as `localcomm` would break the rest of the code in this routine that depends on `localcomm`.

Also, it might be good to stick with the same indentation as elsewhere in this routine.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,240 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      
+      ! Compute the displacements for gathering all the local_block_id_arr to global_block_id_arr
+      global_start(1) = 0
+         do i=2,dminfo % nprocs
```
The indentation seems off here.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,240 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+         end if
+         write(blockFilename,'(a,i0)') trim(partitionFilePrefix), dminfo % nProcs
+         
+         call mpas_log_write('Writing out Scotch Graph paritioning data to '//trim(blockFilename))
```
`paritioning` -> `partitioning`

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,465 @@
+
+    ! Dummy type declaration for SCOTCH distributed graph
+    ! Member ptr is a pointer to the SCOTCH_Dgraph C structure
+    type, bind(c) :: scotchm_dgraph
```
I don't think this `type` needs to be `bind(c)` since instances of this type are never directly passed to C code; only the `type(c_ptr)` member is passed to C code as far as I can tell, and a `type(c_ptr)` is by definition interoperable with a C data pointer.

Similarly for the `scotchm_strat` derived type.

abishekg7 replied:
Thanks for the catch!

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,465 @@
+
+        implicit none
+        ! Arguments
+
```
Was this blank line intended to go above the `! Arguments` comment line?

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,295 @@
+#include <unistd.h>
+#include <string.h>
+#include <stdio.h>
+#include <mpi.h>
```
From the perspective of the C compiler, MPI is a non-system library just like PT-Scotch, so I think we've typically used quotes when including `mpi.h`; i.e.,
```c
#include "mpi.h"
```

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,295 @@
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <string.h>
```
You might want to double check, but it looks like the following headers don't need to be included:
- `stddef.h`
- `sys/stat.h`
- `unistd.h`
- `string.h`

abishekg7 replied:
Checked that they can be removed.

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,295 @@
+
+	*dgraph_ptr = (SCOTCH_Dgraph *) malloc(sizeof (SCOTCH_Dgraph));
+
+	err = SCOTCH_dgraphInit(*dgraph_ptr, comm);
```
Just a suggestion (and it looks like you've done this elsewhere), if you'd like to avoid having to declare the local variable `err`, you could directly return the result of the call to `SCOTCH_dgraphInit`.

abishekg7 replied:
Got it. I was iterating through some designs, and perhaps it can be removed now.

mgduda commented on src/framework/ptscotch_interface.c:
```diff
@@ -0,0 +1,295 @@
+							 edgelocsiz,
+							 edgeloctab,
+							 edgegsttab,
+							 edloloctab);
```
Out of curiosity, was there a reason for using one line for each argument here? Other function calls in this file list multiple arguments on the same line.

abishekg7 replied:
I might have originally wanted a comment after each argument, but it wasn't necessary. I can change this.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,240 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      character (len=*), intent(out) :: blockFilename !< Output: Block decomposition file name
+
+      integer, dimension(:), pointer :: global_start
+      integer, dimension(:), allocatable :: local_cell_list
```
The `gfortran` compiler with `-Wall` reports `local_cell_list` as being declared but never used.

abishekg7 replied:
I can comment this declaration line out as `local_cell_list ` could possibly be needed at a later point if we decide to use the partition data directly from the redistributed graph, instead of writing out to a partition file and then reading from it.

mgduda replied:
In this case, it might improve readability to add a space between the `!` character and the beginning of the variable declaration.

abishekg7 replied:
Done

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,465 @@
+        ! Arguments
+
+        integer(c_int) :: dummy_int
+        integer(c_size_t) :: dummy_size
```
It doesn't look like `dummy_size` is used in this routine.

mgduda commented on src/framework/mpas_ptscotch_interface.F:
```diff
@@ -0,0 +1,465 @@
+
+       if (scotchm_get_intsize() /= c_sizeof(dummy_int)) then
+          call mpas_log_write("Error: Scotch SCOTCH_Num size does not match MPAS integer size \n &
+                                Please build Scotch with 32 bit integers", MPAS_LOG_CRIT)
```
The `gfortran` compiler warns about this continued character constant:
```
mpas_ptscotch_interface.F:56:33:

   56 |                                 Please build Scotch with 32 bit integers", MPAS_LOG_CRIT)
      |                                 1
Warning: Missing '&' in continued character constant at (1) [-Wampersand]
```

mgduda reviewed (changes requested):
I think we're getting really close, and I've just left a few more clean-up comments.

abishekg7 commented:
@mgduda thanks for the comments! I've addressed most of them, except for the `local_cell_list` declaration.

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -95,53 +103,86 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+               useScotch = .true.
+#else
               call mpas_log_write('Could not open block decomposition file for $i blocks.', MPAS_LOG_ERR, intArgs=(/dminfo % total_blocks/) )
               call mpas_log_write('Filename: '//trim(filename), MPAS_LOG_CRIT)
```
Could we add one space of indentation for these calls to `mpas_log_write` so their indentation matches the lines around them?

mgduda commented on src/framework/mpas_block_decomp.F:
```diff
@@ -285,8 +326,240 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
 
+
+
+
```
It looks like the rest of the routines in this module use just one blank line between routines. For consistency, should we do the same here?

mgduda commented on Makefile:
```diff
@@ -759,6 +759,16 @@ endif
+	CPPINCLUDES += $(SCOTCH_FCINCLUDES)
+	LIBS += $(SCOTCH_LIBS)
+	override CPPFLAGS += $(SCOTCH_FFLAGS)
+endif
```
Since it's only C code that needs to interface directly with the PT-Scotch library, we might be able to simplify this block as:
```make
ifneq "$(SCOTCH)" ""
        SCOTCH_INCLUDES += -I$(SCOTCH)/include
        SCOTCH_LIBS += -L$(SCOTCH)/lib64 -lptscotch -lscotch -lptscotcherr -lm
        SCOTCH_FLAGS = -DMPAS_SCOTCH

        CPPINCLUDES += $(SCOTCH_INCLUDES)
        LIBS += $(SCOTCH_LIBS)
        override CPPFLAGS += $(SCOTCH_FLAGS)
endif
```
(Note that the `FC` in `FCINCLUDES` was meant -- as far as I recall -- to refer to "Fortran Compiler"; similarly, `FFLAGS` was "Fortran Flags". So, the suggestion above removes `FC` and `F` prefixes).

abishekg7 commented:
Addressed the remaining review comments, and following the C-indentation style found in `smiol.c`

mgduda commented:
There are currently quite a few messages that are written to the log files when PT-Scotch is used to create a new graph partitioning:
```
 No existing block decomposition file found, invoking Scotch.
 Scotch SCOTCH_Num size is compatible with MPAS integer size
 Successfully initialized distributed Scotch graph
 Successfully initialized distributed Scotch graph
 Before SCOTCH distributed graph build. nLocVertices =2729  nLocEdges= 16359
 Successfully built distributed Scotch graph
 Successfully checked distributed Scotch graph
 Successfully initialized Scotch strategy
 Successfully partitioned distributed Scotch graph
 Successfully redistributed Scotch graph
 Writing out Scotch Graph partitioning data to colorado.graph.info.part.256
 Scotch partition successful
 After Scotch decomposition, attempting to read block decomposition file: colorado.graph.info.part.256
 mpas_block_decomp_cells_for_proc successful
```
I suspect some of these were primarily added to aid in development, and now that the functionality is in good shape, they could be removed.

@abishekg7 Could you go through and remove some of these log messages?

mgduda commented:
Near the beginning of the log file, we write out a summary of compile-time options:
```
 Compile-time options:
   Build target: gnu
   OpenMP support: no
   OpenACC support: no
   Default real precision: single
   Compiler flags: optimize
   I/O layer: SMIOL
   MUSICA support: no
```
As we've done for the MUSICA library, I think it would be helpful to add a similar message for the PT-Scotch library. This could be done in the [`mpas_framework_report_settings`](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.3.1/src/framework/mpas_framework.F#L190-L201) routine in `mpas_framework.F`,

abishekg7 commented:
> There are currently quite a few messages that are written to the log files when PT-Scotch is used to create a new graph partitioning:
> 
> ```
>  No existing block decomposition file found, invoking Scotch.
>  Scotch SCOTCH_Num size is compatible with MPAS integer size
>  Successfully initialized distributed Scotch graph
>  Successfully initialized distributed Scotch graph
>  Before SCOTCH distributed graph build. nLocVertices =2729  nLocEdges= 16359
>  Successfully built distributed Scotch graph
>  Successfully checked distributed Scotch graph
>  Successfully initialized Scotch strategy
>  Successfully partitioned distributed Scotch graph
>  Successfully redistributed Scotch graph
>  Writing out Scotch Graph partitioning data to colorado.graph.info.part.256
>  Scotch partition successful
>  After Scotch decomposition, attempting to read block decomposition file: colorado.graph.info.part.256
>  mpas_block_decomp_cells_for_proc successful
> ```
> 
> I suspect some of these were primarily added to aid in development, and now that the functionality is in good shape, they could be removed.
> 
> @abishekg7 Could you go through and remove some of these log messages?

I can retain these messages for the successful case, and remove the rest. 

```
No existing block decomposition file found, invoking Scotch.
Successfully built distributed Scotch graph
Successfully partitioned distributed Scotch graph
Writing out Scotch Graph partitioning data to colorado.graph.info.part.256
```

abishekg7 committed:
Adding support for linking with the SCOTCH graph partitioning library

This commit adds support for linking MPAS with the SCOTCH graph partitioning
library. This capability is intended as a first step towards MPAS cores being
able to generate graph partitions during runtime, by invoking the distributed
graph partitioning algorithms provided by PT-SCOTCH.

In order for MPAS to be able to link with the SCOTCH library, the SCOTCH
environment variable must be set to point to the SCOTCH installation path
prior to build MPAS.

export SCOTCH=/path/to/scotch/installation
make gnu CORE=atmosphere

If the SCOTCH env variable has been set, but it does not point to the correct
SCOTCH installation directory, then the Scotch C test in the Makefile would
fail, yielding the following message:

Could not build a simple C program with Scotch.

If MPAS has been successfully linked with the SCOTCH library, the build summary
will include the message:

MPAS has been linked with the Scotch graph partitioning library.

Furthermore, the log file at model runtime will also include the following
message near the top:

SCOTCH support: yes

Conversely, if MPAS has not been linked with SCOTCH, then appropriate messages
are displayed in the aforementioned locations.

abishekg7 committed:
Adding source files that enable Fortran-C interfaces from MPAS to SCOTCH

This commit adds a C and Fortran source file to the src/framework directory,
and builds on the previous commit to provide a set of routines which enable
various MPAS cores to invoke the SCOTCH library to do the following tasks:

- Initialize, build and check a distributed SCOTCH graph
- Partition and redistribute the distributed graphs
- Query distributed graphs
- Initialize SCOTCH graph mapping strategies

Together these routines enable the MPAS framework to generate online graph
partitioning. This commit also includes modifications to the src/framework
Makefile to build and C and Fortran interfaces.

To avoid unnecessary references to the SCOTCH fortran library, this commit
introduces two derived types, scotchm_dgraph and scotchm_strat in the
Fortran mpas_ptscotch_interface. These two derived types hold C pointers to
memory locations of type SCOTCH_Dgraph and SCOTCH_Strat respectively, which
are allocated in the C-interface.

abishekg7 committed:
Adding online graph partitioning routine to mpas_block_decomp.F

This commit introduces a subroutine mpas_block_decomp_scotch in
src/framework/mpas_block_decomp.F, which calls the routines introduced in the
previous commit to invoke Scotch libraries.

- The routine initializes and builds PT-SCOTCH distributed graphs from the
  adjacency data available in the partial_global_graph on each MPI task.

- After checking the distributed graph for consistency, the SCOTCH graph
  mapping strategy is initialized, followed by the graph partitioning.
  Following a successful redistribution, the graph must be redistributed so
  that each MPI task is aware of all the vertices that it owns.

- Finally, the routine writes out the generated graph partitioning data to
  disk so that it may be reused for the next model run.

abishekg7 committed:
Enabling online graph partitioning via SCOTCH when partition file not found

This commit modifies the existing logic in mpas_block_decomp_cells_for_proc
routine inside src/framework/mpas_block_decomp.F to call the
mpas_block_decomp_scotch routine, when all of the following are satisfied:

- The MPAS core has been built with SCOTCH support
- The routine is unable to find a relevant graph partition file on disk

As described in the previous commit, the call to mpas_block_decomp_scotch
begins the process of online graph partitioning, resulting in the generated
partition file being written out to disk. Subsequently, the routine
mpas_block_decomp_cells_for_proc attempts to read this newly-generated
partition file, and if successful, proceeds with the rest of initialization.

abishekg7 committed:
Adding new timers to mpas_block_decomp.F

This commit introduces new timers to the mpas_block_decomp_cells_for_proc
and the mpas_block_decomp_scotch routines in src/framework/mpas_block_decomp.F.
These timers may be useful in tracking the time spent in online graph
partitioning or in reading the graph partition files from disk and in
collective MPI calls.

mgduda reviewed (approved):
Nice work -- this looks good to me!
